test/cmd/apply.sh

```diff
@@ -27,75 +27,75 @@ run_kubectl_apply_tests() {
   kube::log::status "Testing kubectl apply"
   ## kubectl apply should create the resource that doesn't exist yet
   # Pre-Condition: no POD exists
-  kube::test::get_object_assert pods "{{range.items}}{{$id_field}}:{{end}}" ''
+  kube::test::get_object_assert pods "{{range.items}}{{${id_field:?}}}:{{end}}" ''
   # Command: apply a pod "test-pod" (doesn't exist) should create this pod
-  kubectl apply -f hack/testdata/pod.yaml "${kube_flags[@]}"
+  kubectl apply -f hack/testdata/pod.yaml "${kube_flags[@]:?}"
   # Post-Condition: pod "test-pod" is created
-  kube::test::get_object_assert 'pods test-pod' "{{${labels_field}.name}}" 'test-pod-label'
+  kube::test::get_object_assert 'pods test-pod' "{{${labels_field:?}.name}}" 'test-pod-label'
   # Post-Condition: pod "test-pod" has configuration annotation
-  [[ "$(kubectl get pods test-pod -o yaml "${kube_flags[@]}" | grep kubectl.kubernetes.io/last-applied-configuration)" ]]
+  kubectl get pods test-pod -o yaml "${kube_flags[@]:?}" | grep -q kubectl.kubernetes.io/last-applied-configuration
   # Clean up
-  kubectl delete pods test-pod "${kube_flags[@]}"
+  kubectl delete pods test-pod "${kube_flags[@]:?}"
 
 
   ## kubectl apply should be able to clear defaulted fields.
   # Pre-Condition: no deployment exists
-  kube::test::get_object_assert deployments "{{range.items}}{{$id_field}}:{{end}}" ''
+  kube::test::get_object_assert deployments "{{range.items}}{{${id_field:?}}}:{{end}}" ''
   # Command: apply a deployment "test-deployment-retainkeys" (doesn't exist) should create this deployment
-  kubectl apply -f hack/testdata/retainKeys/deployment/deployment-before.yaml "${kube_flags[@]}"
+  kubectl apply -f hack/testdata/retainKeys/deployment/deployment-before.yaml "${kube_flags[@]:?}"
   # Post-Condition: deployment "test-deployment-retainkeys" created
-  kube::test::get_object_assert deployments "{{range.items}}{{$id_field}}{{end}}" 'test-deployment-retainkeys'
+  kube::test::get_object_assert deployments "{{range.items}}{{${id_field:?}}}{{end}}" 'test-deployment-retainkeys'
   # Post-Condition: deployment "test-deployment-retainkeys" has defaulted fields
-  [[ "$(kubectl get deployments test-deployment-retainkeys -o yaml "${kube_flags[@]}" | grep RollingUpdate)" ]]
-  [[ "$(kubectl get deployments test-deployment-retainkeys -o yaml "${kube_flags[@]}" | grep maxSurge)" ]]
-  [[ "$(kubectl get deployments test-deployment-retainkeys -o yaml "${kube_flags[@]}" | grep maxUnavailable)" ]]
-  [[ "$(kubectl get deployments test-deployment-retainkeys -o yaml "${kube_flags[@]}" | grep emptyDir)" ]]
+  kubectl get deployments test-deployment-retainkeys -o yaml "${kube_flags[@]:?}" | grep -q RollingUpdate
```

Comment:
`grep -q` will cause pipe errors when used this way... see https://github.com/kubernetes/kubernetes/pull/57805

Reply:
dug into this a bit more, bash makes a tempfile, writes the herestring contents to it, and then execs the command with the tempfile as stdin.

`strace -e execve,open,dup2,pipe,read,write -f bash -c 'cat <<<"test"'`